### PR TITLE
Introduce volumemgr

### PR DESCRIPTION
**Pull request by eriknordmark**
This is refactoring extisting functionality while making it easier to introduce new types of volumes (such as named Volumes, volumes created from scratch, etc)

And there are some loose ends around HandleCert - the image signifing certificates (which are hard to test since they are currently disabled in the controller). But the intent is to complete that logic to the same state as before.

Including flexible Key scheme for VolumeConfig/VolumeStatus

Move handling of downloader, verifier, image certs, and /persist/img to volumemgr
Add explicit support for purgeCounter to recreate volumes
Container prepare split between volumemgr and domainmgr using json file to pass info

Signed-off-by: eriknordmark <erik@zededa.com>

### CODE REVIEW

**Comment by deitch** [comment]
Is there a write-up included in this PR describing the flows? We had that diagram in the Google Doc we shared which described which part does what. Can we get that in here, so there is a "before-and-after" (or just "after") to look at?

The other part that threw me off is, what moves from domainmgr to volumemgr? It looks like the `ctrPrepare` and even `ctrRun` and `ctrStop` move to volumemgr. Is that correct? What are the responsibilities of each once this is done?

**Comment by eriknordmark** [comment]
> Is there a write-up included in this PR describing the flows? We had that diagram in the Google Doc we shared which described which part does what. Can we get that in here, so there is a "before-and-after" (or just "after") to look at?

The flow at what level of detail?
It would make sense to add the handling of purge to the google doc. The rest of the flow is the same when it comes to download and verify except that domaingr has been inserted between baseosmgr/zedmanger and verifier/downloader.

I can describe the new flow in the document (but a delta against the old doesn't have very long shelf life.)

> The other part that threw me off is, what moves from domainmgr to volumemgr? It looks like the `ctrPrepare` and even `ctrRun` and `ctrStop` move to volumemgr. Is that correct? What are the responsibilities of each once this is done?

I didn't realize those existed (since they are empty and unused). They belong in domainmgr. I'll move the empty functions there.

**Comment by deitch** [comment]
> The flow at what level of detail?

I meant something simple, like:

controller --> zed manager ---> volumemgr (handles requests)
                                                           |
                                                           |
                                                           domainmgr

Just a clear understanding of who owns what.

Yeah, a delta has a low shelf life.

**Comment by rvs** [comment]
It is really awesome to see this come in @eriknordmark -- a couple of quick observations:
   * should we move the remaining code from containerd.go into domainmanager? There's nothing containerd related in those two functions -- not even a containerd API import is left
   * the biggest amount of questions I have is about the modeling that is happening in VolumeConfig type and how it would dovetail in us evolving this further https://github.com/lf-edge/eve/pull/833/files#diff-5a4e94c5ba8bf064f995ddf3081ed1f9R87 -- but this is going forward, since it looks like this PR simply put all the current implementation together -- so we basically get today's behavior
   * I am a little big confused as to how the new flow between domainmanager, zedmagaer and volumemanegr is structured. Specifically how and where DomainConfig gets filled and what are the expectations. Can you please comment on that?

Basically, the last question is really about -- who is subscribed to whom publishing what.

Finally, I'd love to see this merged ASAP -- what kind of tests/etc. would currently block that?

**Comment by eriknordmark** [comment]
@deitch @rvs documentation in https://github.com/lf-edge/eve/pull/833/commits/b8374becd72785a58191dae28b7ffc1a664dbac9

**Comment by deitch** [comment]
Thanks for adding the docs, and moving those container funcs out. It looks good to me. I wouldn't mind seeing a flow description like we put in that Google Doc you and I discussed offline @eriknordmark, but I wouldn't hold it up for that.

Something like:

There is a distinction between volumes and images:

* Images are immutable read/only blobs of data
* Volumes are read/write or read/only mount points for one or more apps that are driven from images

The current volumemgr design is as follows (not all details are included):

* zedmanager
   * Receives request for app with at least one volume, or direct volumes. 
   * Requests image hash from resolver
   * Requests volumes from volumemgr:
      * Checks if the required volumes already exist; if they do, simply return
      * Checks if the immutable image for the volume exists. If not:
      * Requests image download from downloadmgr:
         * Downloads image
      * Requests image verification from verifier
         * Verifies the image
      * If container image: Loads the image into the containerd store
      * Creates the volume based on the image
      * Reports to zedmanager
   * Request domain start from domainmgr
      * Starts domain
   * Complete

**Comment by eriknordmark** [comment]
@deitch The zedmanager part of the description really belongs in a zedmanager.md file.
I'll update the volumemgr about the volume vs. image
